### PR TITLE
Assignment 8 Doug Doner

### PR DESCRIPTION
**Pull request by dougdoner**

### CODE REVIEW

File: Assignment.Tests/SampleDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+using System.Text;
+using System.Linq;
+
+namespace Assignment.Tests
+{
+    [TestClass]
+    public class SampleDataTests
+    {
+        [TestMethod]
+        [DataRow("C:\\Users\\me\\Documents\\repos\\EWU-CSCD371-2019-Fall\\Assignment\\People.csv")]
```
Comment (ashermancinelli) [inline Assignment.Tests/SampleDataTests.cs]: Consider adding a reference item with a relative path to the csv file - an absolute path is not portable and will always fail the on Azure (or in any other dir/any other machine).

Comment (dougdoner) [reply Assignment.Tests/SampleDataTests.cs]: Thanks for pointing that out. I had trouble at first understanding where relative paths are executed, but it's apparently relative to the executable (in this case buried in the Assignment.tests bin folder).

I've put in an initial fix in 7369fff.

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,96 @@
...
     public class SampleData : ISampleData
     {
         // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
+        public IEnumerable<string> CsvRows
+        {
+            get
+            {
+                List<string> result = new List<string>();
+                using (var fileReader = new StreamReader(@$".\People.csv"))
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: `File.ReadAllTest()` would be so much easier. :)

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,96 @@
...
 
         // 2.
-        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() 
-            => throw new NotImplementedException();
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            SampleData sampleData = new SampleData();
+            IEnumerable<IPerson> personCollection = sampleData.People;
+
+            IEnumerable<string> statesList = personCollection.Select(item => item.Address.State).Distinct<string>().OrderBy(s => s);
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: If the file was missing an address here, `Address` could be null. (I suspect you would be warned on this if Nullable is enabled.)

File: Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,96 @@
...
+            return statesList;
+        }
 
         // 3.
         public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+        {
+            SampleData sample = new SampleData();
+            IEnumerable<string> stateCollection = sample.GetUniqueSortedListOfStatesGivenCsvRows();
+            string[] stateStringArray = stateCollection.ToArray<string>();
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Join takes an IEnumerable<string> so you don't need to call `ToArray()`.

**Comment by MarkMichaelis** [comment]
##  (dougdoner)

## Assignment8

| Build/Compile (Build Succeeded: 0 Error(s), 23 Warning(s))| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5**


| Assignment  | **10** |
| --------------------- | ------ |
| 1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔ <br/>- Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔ <br/> - Using LINQ, skip the first row in the `People.csv`. ✔ <br/> - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔  | ✔ |
| 2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔ <br/> Use `ISampleData.CsvRows` for your data source.❌ <br/>- Don't forget the list should be unique. ✔ <br/>- Sort the list alphabetically ✔ <br/>- Include a test that leverages a hard coded list of Spokane based addresses. ❌ <br/>- Include a test that uses LINQ to verify the data is sorted correctly (do not use a hard coded list). ✔  | ✔ |
| 3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` for your data source. ✔ <br/>- Given the array, consider using `string.Join` to combine the list into a single string.  | ✔ |
| 4.Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔ <br/>- Use `ISampleData.CsvRows` as the source of the data. ✔ <br/>- Sort the list by State, City, Zip. (Sort the addresses first then select). ❌ <br/>- Be sure that `Person.Address` is also populated. ✔ <br/>- Adding null validation to all the `Person` and `Address` properties is **optional**. <br/>- Consider using `ISampleData.CsvRows` in your test to verify your results.  | ✔ |
| 5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔ <br/>- Use `ISampleData.People` for your data source. ✔  | ✔ |
| 6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use the `people` parameter from `ISampleData.GetUniqueListOfStates` for your data source. ✔ <br/>- At a minimum, use `System.Linq.Enumerable.Aggregate` LINQ method to create your result.✔ <br/>- Don't forget the list should be unique. ✔ <br/>- It is recommended that at a minimum you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result.  | ✔ |


| Project Analysis |   |
| -----------------| - |
| Disabled Warnings | **✔**
| Nullable Enabled: | **❌** |
| CodeReview Performed | **✔** |
| Appropriately handle all warnings | **❌** |
| Add the code analysis packages and appropriately handle all warnings* |  **✔** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Assignment | Disable | Assignment8.sln |  | | 4 |
| Assignment.Tests | Disable | Assignment8.sln |  |  | 4 |


| Testing | **20** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=10, Passed=10, Failed=0, Skipped=0 | **97%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **5** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write awesome, well tested code that follows all the guidelines outlined in the book. | ❌ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance. | ❌ |
| Implement all homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `ISampleData`). Refactor both your SampleData and SampleDataAsync classes so there is minimal duplication.  This includes refactoring your tests so a significant amount of test code can be re-used.| ❌ |

**Pull request closed without merging**